cjwkernel/pandas/types.py

```diff
@@ -770,75 +853,119 @@ def coerce(
-                value["columns"] = _infer_columns(
-                    dataframe, column_formats, try_fallback_columns
+                errors = (
+                    [(old_error.pop("error"), old_error.pop("quick_fixes", []))]
+                    if old_error
+                    else []
                 )
-            except KeyError:
-                pass
 
-            try:
-                return cls(dataframe=dataframe, **value)
-            except TypeError as err:
-                raise ValueError(
-                    (
-                        "ProcessResult input must only contain {dataframe, "
-                        "error, json, quick_fixes, column_formats} keys"
+                try:
+                    errors = errors + _coerce_to_process_result_error(
+                        value.pop("errors")
+                    )
+                except KeyError:
+                    pass
+
+                dataframe = value.pop("dataframe", pd.DataFrame())
+                validate_dataframe(dataframe)
+
+                try:
+                    column_formats = value.pop("column_formats")
+                    value["columns"] = _infer_columns(
+                        dataframe, column_formats, try_fallback_columns
                     )
-                ) from err
+                except KeyError:
+                    pass
+
+                try:
+                    return cls(dataframe=dataframe, errors=errors, **value)
+                except TypeError as err:
+                    raise ValueError(
+                        (
+                            "ProcessResult input must only contain {dataframe, "
+                            "error, errors, json, quick_fixes, column_formats} keys"
+                        )
+                    ) from err
         elif isinstance(value, tuple):
             if len(value) == 2:
-                dataframe, error = value
-                if dataframe is None:
-                    dataframe = pd.DataFrame()
-                if error is None:
-                    error = ""
-                if not isinstance(dataframe, pd.DataFrame) or not isinstance(
-                    error, str
-                ):
-                    return cls(
-                        error=(
-                            (
-                                "There is a bug in this module: expected "
-                                "(DataFrame, str) return type, got (%s,%s)"
+                if isinstance(value[0], pd.DataFrame) or value[0] is None:
+                    dataframe, error = value
+                    if dataframe is None:
+                        dataframe = pd.DataFrame()
+                    try:
+                        errors = (
+                            []
+                            if error is None
+                            else _coerce_to_process_result_error(error)
+                        )
+                    except TypeError:
+                        errors = None
```

Comment:
I wouldn't prefer hiding the error either, but looks like `ProcessResult.coerce` never wants to raise when its argument is a tuple; it's also tested this way.

Reply:
Ah -- whoops, you're right, I forgot the contract. Indeed, we aren't meant to raise a `ValueError`: we're meant to return a `ProcessResult` with an error message.

My comment still holds: I think it would be nice if the error-`ProcessResult` (returned two lines down) included the `TypeError` text.

(And I'm still not 100% sure that's what we want. I'm comfortable merging this as-is and revisiting later if a user asks.)

Reply:
It would be a little tricky: when the `TypeError` is constructed, we don't not know where it may end up. If we construct it with a string as error, we will not be able to convert the error to `I18nMessage` in this part of `coerce` (at least without using regexps, which however seem a monstrous idea). If we construct it with an `I18nMessage` as error, we will not have a pretty error message in other parts of `coerce` where the `TypeError` is not caught (or, in fact, caught and reraised as `ValueError`).

Reply:
These messages aren't i18n. While the errors can appear to end users, they are error _codes_ destined for module authors. End users won't understand them, no matter what language they're in.

Reply:
In this case, it's easy to include it. I pushed a commit and I am thus ready for a review (or merge).